File: BE/Core/BookStore.Domain/Entities/Catalog/Book.cs
using BookStore.Domain.Entities.Common;
using BookStore.Domain.Entities.Pricing___Inventory;
using BookStore.Domain.Entities.Pricing_Inventory;
using BookStore.Domain.Entities.Rental;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using BookStore.Domain.ValueObjects;

namespace BookStore.Domain.Entities.Catalog
{
    public class Book
    {
        public Guid Id { get; set; }

        public string Title { get; set; } = null!;           
        public ISBN ISBN { get; set; } = null!;           
        public string? Description { get; set; } = null!;       
        public int PublicationYear { get; set; }             
        public string Language { get; set; } = "vi";          
        public bool IsAvailable { get; set; } = true;         
        public string? Edition { get; set; }                  
        public int PageCount { get; set; }                   

        //  1-n: Mỗi sách thuộc 1 nhà xuất bản
        public Guid PublisherId { get; set; }
        public virtual Publisher Publisher { get; set; } = null!;

        // Foreign Key
        public Guid? BookFormatId { get; set; }

        public virtual BookFormat? BookFormat { get; set; }
        //  n-n: Nhiều tác giả cho 1 sách
        public virtual ICollection<BookAuthor> BookAuthors { get; set; } = new List<BookAuthor>();

        //  n-n: Sách có thể thuộc nhiều thể loại
        public virtual ICollection<BookCategory> BookCategories { get; set; } = new List<BookCategory>();

        //  1-n: Ảnh, file, metadata
        public virtual ICollection<BookImage> Images { get; set; } = new List<BookImage>();
        public virtual ICollection<BookFile> Files { get; set; } = new List<BookFile>();
        public virtual ICollection<BookMetadata> Metadata { get; set; } = new List<BookMetadata>();

        //  1-n: Quan hệ sang module Pricing & Review
        public virtual ICollection<Price> Prices { get; set; } = new List<Price>();
        public virtual StockItem? StockItem { get; set; }            
        public virtual ICollection<Review> Reviews { get; set; } = new List<Review>();
        
        // Review statistics
        public decimal AverageRating { get; set; } = 0;         // Điểm trung bình
        public int TotalReviews { get; set; } = 0;              // Tổng số đánh giá
        
        public virtual ICollection<BookRental> Rentals { get; set; } = new List<BookRental>(); 
    }
}

File: app-frontend/src/types/book.ts
// Book types matching backend DTOs

// Base Book type from BookDto (for list views)
export interface Book {
  id: string; // Guid from backend
  title: string;
  isbn: string;
  publicationYear: number;
  language: string;
  pageCount: number;
  isAvailable: boolean;
  
  // Publisher
  publisherId: string;
  publisherName: string;
  
  // Format
  bookFormatId?: string;
  bookFormatName?: string;
  
  // Authors & Categories (list of names)
  authorNames: string[];
  categoryNames: string[];
  
  // Pricing
  currentPrice?: number;
  discountPrice?: number;
  
  // Stock
  stockQuantity?: number;
  
  // Reviews
  averageRating?: number;
  totalReviews: number;
  
  // For UI display (optional, from images)
  cover?: string;
  coverImages?: string[];
}

// Detailed Book type from BookDetailDto (for detail view)
export interface BookDetail {
  id: string;
  title: string;
  isbn: string;
  description?: string;
  publicationYear: number;
  language: string;
  edition?: string;
  pageCount: number;
  isAvailable: boolean;
  
  publisher: Publisher;
  bookFormat?: BookFormat;
  authors: Author[];
  categories: Category[];
  images: BookImage[];
  files: BookFile[];
  metadata: BookMetadata[];
}

// Supporting types
export interface Publisher {
  id: string;
  name: string;
  description?: string;
  website?: string;
  email?: string;
  phone?: string;
  address?: string;
  country?: string;
  foundedYear?: number;
}

export interface BookFormat {
  id: string;
  name: string;
  description?: string;
}

export interface Author {
  id: string;
  name: string;
  biography?: string;
  birthDate?: string;
  nationality?: string;
  email?: string;
  website?: string;
}

export interface Category {
  id: string;
  name: string;
  description?: string;
  slug?: string;
  parentCategoryId?: string;
}

export interface BookImage {
  id: string;
  bookId: string;
  imageUrl: string;
  isCover: boolean;
  displayOrder: number;
}

export interface BookFile {
  id: string;
  bookId: string;
  fileUrl: string;
  fileType: string;
  fileSize: number;
  uploadDate: string;
}

export interface BookMetadata {
  id: string;
  bookId: string;
  key: string;
  value: string;
}

// API Request/Response types
export interface GetBooksParams {
  pageNumber?: number;
  pageSize?: number;
  keyword?: string;
  categoryId?: string;
  authorId?: string;
  publisherId?: string;
  minPrice?: number;
  maxPrice?: number;
  language?: string;
  sortBy?: string;
  sortDescending?: boolean;
}

export interface BookListResponse {
  items: Book[];
  totalCount: number;
  pageNumber: number;
  pageSize: number;
  totalPages: number;
  hasNextPage: boolean;
  hasPreviousPage: boolean;
}

// Helper function to convert Book to legacy Book format (for compatibility with existing components)
export const toDisplayBook = (book: Book): {
  id: number;
  title: string;
  author: string;
  cover: string;
  color?: string;
  rating?: number;
  reviewCount?: number;
  price?: number;
  description?: string;
  pages?: number;
  language?: string;
  publisher?: string;
} => ({
  id: parseInt(book.id.split('-')[0], 16), // Convert UUID to number for compatibility
  title: book.title,
  author: book.authorNames.join(', '),
  cover: book.cover || 'https://via.placeholder.com/160x240?text=No+Cover',
  rating: book.averageRating,
  reviewCount: book.totalReviews,
  price: book.discountPrice || book.currentPrice,
  pages: book.pageCount,
  language: book.language,
  publisher: book.publisherName,
});

// Helper to convert BookDetail to display format
export const toDisplayBookDetail = (book: BookDetail): any => ({
  id: book.id, // Keep original UUID string for API calls
  title: book.title,
  author: book.authors.map(a => a.name).join(', '),
  cover: book.images.find(img => img.isCover)?.imageUrl || book.images[0]?.imageUrl || 'https://via.placeholder.com/220x320?text=No+Cover',
  coverImages: book.images.sort((a, b) => a.displayOrder - b.displayOrder).map(img => img.imageUrl),
  description: book.description,
  pages: book.pageCount,
  language: book.language,
  publisher: book.publisher.name,
  rating: 0, // TODO: Get from reviews
  // derive price and currency from any attached fields
  price: (book as any).price ?? (book as any).currentPrice ?? (book as any).amount ?? 0,
  currency: (book as any).currency ?? 'VND',
});
